pettingzoo/test/seed_test.py

```diff
@@ -1,109 +1,128 @@
-import hashlib
-import pickle
-import random
-import warnings
-
-import numpy as np
-
-from pettingzoo.utils import parallel_to_aec
-
-
-def hash(val):
-    val = pickle.dumps(val)
-    hasher = hashlib.md5()
-    hasher.update(val)
-    return hasher.hexdigest()
-
-
-def calc_hash(new_env, rand_issue, max_env_iters):
-    cur_hashes = []
-    sampler = random.Random(42)
-    for i in range(3):
-        new_env.reset(seed=i)
-        for j in range(rand_issue + 1):
-            random.randint(0, 1000)
-            np.random.normal(size=100)
-        for agent in new_env.agent_iter(max_env_iters):
-            obs, rew, terminated, truncated, info = new_env.last()
-            if terminated or truncated:
-                action = None
-            elif isinstance(obs, dict) and "action_mask" in obs:
-                action = sampler.choice(np.flatnonzero(obs["action_mask"]).tolist())
-            else:
-                action = new_env.action_space(agent).sample()
-            new_env.step(action)
-            cur_hashes.append(agent)
-            cur_hashes.append(hash_obsevation(obs))
-            cur_hashes.append(float(rew))
-
-    return hash(tuple(cur_hashes))
+from gymnasium.utils.env_checker import data_equivalence
 
 
 def seed_action_spaces(env):
-    if hasattr(env, "possible_agents"):
-        for i, agent in enumerate(env.possible_agents):
+    if hasattr(env, "agents"):
+        for i, agent in enumerate(env.agents):
             env.action_space(agent).seed(42 + i)
 
 
+def seed_observation_spaces(env):
+    if hasattr(env, "agents"):
+        for i, agent in enumerate(env.agents):
+            env.observation_space(agent).seed(42 + i)
+
+
 def check_environment_deterministic(env1, env2, num_cycles):
-    """
-    env1 and env2 should be seeded environments
+    """Check that two AEC environments execute the same way."""
 
-    returns a bool: true if env1 and env2 execute the same way
-    """
+    env1.reset(seed=42)
+    env2.reset(seed=42)
 
-    # seeds action space so that actions are deterministic
+    # seed action spaces to ensure sampled actions are the same
     seed_action_spaces(env1)
     seed_action_spaces(env2)
 
-    num_agents = max(1, len(getattr(env1, "possible_agents", [])))
+    # seed observation spaces to ensure first observation is the same
+    seed_observation_spaces(env1)
+    seed_observation_spaces(env2)
+
+    iter = 0
+    max_env_iters = num_cycles * len(env1.agents)
+
+    for agent1, agent2 in zip(env1.agent_iter(), env2.agent_iter()):
+        assert data_equivalence(agent1, agent2), f"Incorrect agent: {agent1} {agent2}"
+
+        obs1, reward1, termination1, truncation1, info1 = env1.last()
+        obs2, reward2, termination2, truncation2, info2 = env2.last()
+
+        if termination1 or truncation1 or termination2 or truncation2:
+            break
```

Comment:
We probably want to do this after the asserts below.

Reply:
Good call, I’ll put it at the end. I think I copied this structure from shimmy or gymnasium so one of those might also do things slightly in the wrong order.